applications: nrf_desktop: Add motion documentation (#1141)
NordicPlayground/fw-nrfconnect-nrf: applications/nrf_desktop/doc/motion.rst

+------------------+-----------------------------------+               +------------------+------------------+
| :ref:`hids`      | ``hid_report_sent_event``         |               |                  |                  |
+------------------+                                   +               +                  +                  +
| :ref:`usb_state` |                                   |               |                  |                  |
+------------------+-----------------------------------+---------------+------------------+------------------+
| :ref:`hids`      | ``hid_report_subscription_event`` |               |                  |                  |
+------------------+                                   +               +                  +                  +
| :ref:`usb_state` |                                   |               |                  |                  |
+------------------+-----------------------------------+---------------+------------------+------------------+

Configuration
*************

The ``motion`` module selects source of movement using following config options.

>     * ``CONFIG_DESKTOP_MOTION_PMW3360_ENABLE`` - selects gaming grade ``PMW3360`` motion sensor.

Author: fixed

    * ``CONFIG_DESKTOP_MOTION_PAW3212_ENABLE`` - selects ``PAW3212`` motion sensor.
    * ``CONFIG_DESKTOP_MOTION_BUTTONS_ENABLE`` - movement is generated using buttons.
    * ``CONFIG_DESKTOP_MOTION_SIMULATED_ENABLE`` - movement is simulated (controlled from shell).
    * ``CONFIG_DESKTOP_MOTION_NONE`` - module is disabled.

Depending on selected config option different implementation file is used during
a build process.

Motion sensors
==============

Choosing any of the motion sensors (``CONFIG_DESKTOP_MOTION_PMW3360_ENABLE``

> or ``CONFIG_DESKTOP_MOTION_PAW3212_ENABLE``) includes ``src/hw_interface/motion_sensor.c``.

Author: Done

The motion sensor is sampled from the context of a dedicated thread.
The option ``CONFIG_DESKTOP_MOTION_THREAD_STACK_SIZE`` is used to set this
thread's stack size.

The motion sensor default sensitivity and power saving switching times can be
set with following options.
    * ``CONFIG_DESKTOP_MOTION_SENSOR_CPI`` - default CPI.

>     * ``CONFIG_DESKTOP_MOTION_SENSOR_SLEEP1_TIMEOUT_MS`` - sleep 1 mode default switch time.

Author: done

    * ``CONFIG_DESKTOP_MOTION_SENSOR_SLEEP2_TIMEOUT_MS`` - sleep 2 mode default switch time.
    * ``CONFIG_DESKTOP_MOTION_SENSOR_SLEEP3_TIMEOUT_MS`` - sleep 3 mode default switch time.
For more information refer to sensor documentation and Kconfig.

> Buttons

Author: I would like to rework the code underneath. At the moment it uses GPIO directly while I believe it should depend on the buttons event. This is why I provided no text here.

=======

Selecting ``CONFIG_DESKTOP_MOTION_BUTTONS_ENABLE`` includes ``src/hw_interface/motion_buttons.c``.

Simulated
=========

Selecting ``CONFIG_DESKTOP_MOTION_SIMULATED_ENABLE`` includes ``src/hw_interface/motion_simulated.c``.

> The ``motion`` module registers a shell module ``motion_sim`` and links to it

Author: done

> The ``motion`` module registers a shell module ``motion_sim`` and links to it

Author: done

> The ``motion`` module registers a shell module ``motion_sim`` and links to it

Author: done

two commands: ``start`` and ``stop``. When started the module will generate
simulated motion events on the predefined trajectory (tracking an eight-sided

> polygon). The option ``CONFIG_DESKTOP_MOTION_SIMULATED_EDGE_TIME`` sets the

Author: done

edge crossing time while the option ``CONFIG_DESKTOP_MOTION_SIMULATED_SCALE_FACTOR``
scales the size of the polygon.

> 

Author: done, I am not aware of any limitations

Implementation details
**********************

Implementation details describe the ``motion`` module implementation for motion
sensors.

Motion sensor
=============

> The ``motion`` module sample data from the motion sensor using the motion

Author: s/sample/samples

> sensor driver. Depending on the selected sensor name of a device linked

Author: ok

> with the driver changes. The same is true for sensor configuration options.

Author: fixed

The ``nrf_desktop`` application aggregates and translates sensor specific
information to the application abstracts in ``configuration/common/motion_sensor.h``

> header file.

Author: ok

Sampling thread
===============

The ``motion`` module uses a dedicated thread to sample data from the motion
sensor. The reason for using the sampling thread is long time required for
data to be ready. Using a thread also greatly simplifies the module as interaction
with the sensor becomes sequential. It is assumed that data sampling will be
done in the background and so the sampling thread priority is set to 0

> (the highest preemptive thread priority).

Author: done

The sampling thread stays in unready state blocked on a semaphore. The semaphore
is triggered when the motion sensor trigger notifies data is available or
other application event requires the module interaction with the sensor
(e.g. when configuration is submitted from the host).

Sampling pipeline
=================

The ``motion`` module starts in ``STATE_DISABLED`` and after initialization

> enters ``STATE_DISCONNECTED``. When disconnected module is not generating

Reviewer: Add a blank line between these sentences for better structure.

> motion events but motion sensor is sampled to make sure its registers are cleared.

Reviewer: ```suggestion
motion events, but the motion sensor is sampled to make sure its registers are cleared.
```

> Upon connection ``motion`` module switches to ``STATE_IDLE`` in this state

Reviewer: ```suggestion
Upon connection, the ``motion`` module switches to the ``STATE_IDLE`` state, in which
```

> module waits for the motion sensor trigger. When first motion is detected and

Reviewer: ```suggestion
the module waits for the motion sensor trigger.

When a first motion is detected and
```

> the device is connected to host (can transmit the HID data out) module switches

Reviewer: ```suggestion
the device is connected to a host (meaning it can transmit the HID data out), the module switches
```

to ``STATE_FETCHING``, samples the motion data and submits the ``motion_event``.
The ``motion`` module then waits for indication that the ``motion_event`` data
was transmitted to the host. This is done when the module receives the

> ``hid_report_sent_event`` event. At that moment a next motion sampling is

Reviewer: ```suggestion
``hid_report_sent_event`` event. At that moment, a next motion sampling is
```

performed and the next ``motion_event`` sent.

The module continues to sample data until disconnection or when there is no
motion detected. The ``motion`` module assumes no motion when ``10`` consecutive

> samples return zero on both axis.

Reviewer: What happns then? It goes back to the IDLE state?

Author: added

    * ``CONFIG_DESKTOP_MOTION_SIMULATED_SCALE_FACTOR`` - scales the size of the polygon.

The ``stop`` command will cause the module to stop generating :new events.

Implementation details
**********************

This section describes the ``motion`` module implementation for motion
sensors.

Motion sensor
=============

The ``motion`` module samples movement data from the motion sensor using the motion
sensor driver.
The name of the device linked with the driver changes depending on the selected sensor.
The same is true for the names of sensor configuration options.
The ``nrf_desktop`` application aggregates the sensor-specific information and
translates them to the application abstracts in the
``configuration/common/motion_sensor.h`` header file.

Sampling thread
===============

The ``motion`` module uses a dedicated sampling thread to sample data from the motion
sensor. The reason for using the sampling thread is the long time required for
data to be ready. Using a dedicated thread also simplifies the module, because the interaction
with the sensor becomes sequential. The sampling thread priority is set to 0

> (the highest preemptive thread priority) because it is assumed that the data

Reviewer: ```suggestion
(the highest preemptive thread priority), because it is assumed that the data
```

sampling will happen in the background.

The sampling thread stays in the unready state blocked on a semaphore. The semaphore
is triggered when the motion sensor trigger sends a notification that the data is available or

> other application event requires the module interaction with the sensor

Reviewer: ```suggestion
when other application event requires the module interaction with the sensor
```

(for example, when configuration is submitted from the host).

Sampling pipeline
=================

The ``motion`` module starts in ``STATE_DISABLED`` and after initialization
enters ``STATE_DISCONNECTED``. When disconnected, the module is not generating
motion events, but the motion sensor is sampled to make sure its registers are cleared.

Upon connection, the ``motion`` module switches to the ``STATE_IDLE`` state, in which
the module waits for the motion sensor trigger.

When a first motion is detected and
the device is connected to a host (meaning it can transmit the HID data out), the module switches
to ``STATE_FETCHING``, samples the motion data and submits the ``motion_event``.
The ``motion`` module then waits for indication that the ``motion_event`` data
was transmitted to the host. This is done when the module receives the
``hid_report_sent_event`` event. At that moment, a next motion sampling is
performed and the next ``motion_event`` sent.

The module continues to sample data until disconnection or when there is no
motion detected. The ``motion`` module assumes no motion when ``10`` consecutive

> samples return zero on both axis. In such case the module will switch back to

Reviewer: ```suggestion
samples return zero on both axis. In such case, the module will switch back to
```
